Docs: Update README for Benchmark C++ Tool and Benchmark Python Tool (#11961)
openvinotoolkit/openvino: samples/cpp/benchmark_app/README.md

# Benchmark C++ Tool {#openvino_inference_engine_samples_benchmark_app_README}

This page demonstrates how to use the Benchmark C++ Tool to estimate deep learning inference performance on supported devices.

> **NOTE**: This page describes usage of the C++ implementation of the Benchmark Tool. For the Python implementation, refer to the [Benchmark Python Tool](../../../tools/benchmark_tool/README.md) page. The Python version is recommended for benchmarking models that will be used in Python applications, and the C++ version is recommended for benchmarking models that will be used in C++ applications. Both tools have a similar command interface and backend.
		

## Basic Usage

> To use the C++ benchmark_app, you must first build it following the [Build the Sample Applications](../../../docs/OV_Runtime_UG/Samples_Overview.md#build-the-sample-applications) instructions and then set up paths and environment variables by following the [Get Ready for Running the Sample Applications](../../../docs/OV_Runtime_UG/Samples_Overview.md#get-ready-for-running-the-sample-applications) instructions. Navigate to the directory where the benchmark_app C++ sample binary was built.

Reviewer: ```suggestion
To use the C++ benchmark_app, you must first build it following the [Build the Sample Applications](../../../docs/OV_Runtime_UG/Samples_Overview.md) instructions and then set up paths and environment variables by following the [Get Ready for Running the Sample Applications](../../../docs/OV_Runtime_UG/Samples_Overview.md) instructions. Navigate to the directory where the benchmark_app C++ sample binary was built.
```

> **NOTE**: If you installed OpenVINO Runtime using PyPI or Anaconda Cloud, only the [Benchmark Python Tool](../../../tools/benchmark_tool/README.md) is available, and you should follow the usage instructions on that page instead.

> The benchmarking application works with models in the OpenVINO IR (`model.xml` and `model.bin`) and ONNX (`model.onnx`) formats. Make sure to [convert your models](../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md) if necessary.

Reviewer 2: Why don't we talk about Paddle?
@helena-intel @kblaszczak-intel Honestly benchmark app supports all framework types which are supported in openvino. If we always will enumerate supported frameworks, we have to update a lot of documentation in case of introducing new framework support.
Supported frameworks should be enumerated on the dedicated page, I guess.

Reviewer 3: PaddlePaddle is currently only supported for 10 selected models. We should not give the impression that our PaddlePaddle support is similar to ONNX at the moment. For now, it would be useful to add "and other supported frameworks" and link to a friendly page that explains what exactly is supported and what the known limitations are.

Reviewer 2: The main message was that we shouldn't enumerate supported frameworks in the documentation for each tools. maybe will be better to have one common place in our documentation and references from different components to this place.

Reviewer: I hope this solution is a proper common ground. It mentions PdPd and states its limitations. 
I hope we'll get to the point where we don't need to mention all of them, just state "natively supported frameworks" but at this stage, this should suffice.

> The benchmarking application works with models in the OpenVINO IR (`model.xml` and `model.bin`) and ONNX (`model.onnx`) formats. Make sure to [convert your models](../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md) if necessary.

Reviewer: ```suggestion
The benchmarking application works with models in the OpenVINO IR, ONNX, and PaddlePaddle (currently, for a limited list of supported models) formats. Make sure to [convert your models](../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md) if necessary.
```

> The benchmarking application works with OpenVINO models in IR format (`model.xml` and `model.bin`) or ONNX format (`model.onnx`). You can use it with your own models or pre-trained [public](@ref omz_models_group_public) or [Intel](@ref omz_models_group_intel) models from the Open Model Zoo, which can be downloaded using the [Model Downloader](@ref omz_tools_downloader). Before running the tool, make sure the model is converted to OpenVINO format using the [Model Optimizer](../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md) tool or is in ONNX format.

Reviewer 3: My suggestion would be to remove the OMZ part from this sentence because in this context this may add some confusion. As a general rule, for OMZ models you should use Model Converter, not MO. And intel models are already in IR format. But if we add all that it is too much information. From my perspective the part from "You can use it with your own models..." could be deleted, but if we want to keep it, can it be just a link to OMZ with if needed a note that public models must be converted with omz_converter (linking to the documentation page about that)?

> The benchmarking application works with OpenVINO models in IR format (`model.xml` and `model.bin`) or ONNX format (`model.onnx`). You can use it with your own models or pre-trained ones. You can choose among both [public](@ref omz_models_group_public) and [Intel](@ref omz_models_group_intel) models from the Open Model Zoo, which you can download using [Model Downloader](@ref omz_tools_downloader). Before running the tool, make sure the model is either converted to the OpenVINO format, using [Model Optimizer](../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md), or is in the ONNX format.

Reviewer: ```suggestion
The benchmarking application works with models in the OpenVINO IR (`model.xml` and `model.bin`) and ONNX (`model.onnx`) formats. Make sure to [convert your models] (../../../docs/MO_DG/Deep_Learning_Model_Optimizer_DevGuide.md) if necessary. 
```

> To run benchmarking with default options on a model, use the following command. 

Reviewer: ```suggestion
To run benchmarking with default options on a model, use the following command:
```

```
./benchmark_app -m model.xml
```

By default, the application will load the specified model onto the CPU and perform inferencing on batches of randomly-generated data inputs for 60 seconds. As it loads, it prints information about benchmark parameters. When benchmarking is completed, it reports the minimum, average, and maximum inferencing latency and average the throughput.

You may be able to improve benchmark results beyond the default configuration by configuring some of the execution parameters for your model. For example, you can use "throughput" or "latency" performance hints to optimize the runtime for higher FPS or reduced inferencing time. Read on to learn more about the configuration options available with benchmark_app.

## Configuration Options
The benchmark app provides various options for configuring execution parameters. This section covers key configuration options for easily tuning benchmarking to achieve better performance on your device. A list of all configuration options is given in the [Advanced Usage](#advanced-usage) section.

### Performance hints: latency and throughput

> The benchmark app allows users to provide high-level "performance hints" for setting latency-focused or throughput-focused inference modes. This hint causes the runtime to automatically adjust runtime parameters, such as number of processing streams and inference batch size, to prioritize for reduced latency or high throughput.

Reviewer: ```suggestion
The benchmark app allows users to provide high-level "performance hints" for setting latency-focused or throughput-focused inference modes. This hint causes the runtime to automatically adjust runtime parameters, such as the number of processing streams and inference batch size, to prioritize for reduced latency or high throughput.
```

The performance hints do not require any device-specific settings and they are completely portable between devices. Parameters are automatically configured based on whichever device is being used. This allows users to easily port applications between hardware targets without having to re-determine the best runtime parameters for the new device.

> The performance hint is set by using `-hint latency` or `-hint throughput` when running benchmark_app:

Reviewer: ```suggestion
If not specified, throughput is used as the default. To set the hint explicitly, use `-hint latency` or `-hint throughput` when running benchmark_app:
```

When benchmark_app is run with `-hint throughput`, it automatically sets the inference batch size to fill up all the memory available. It also maximizes the number of parallel inference requests to utilize all the threads available on the device.

For more information on performance hints, see the [High-level Performance Hints](../../../docs/OV_Runtime_UG/performance_hints.md) page. For more details on optimal runtime configurations and how they are automatically determined using performance hints, see [Runtime Inference Optimizations](../../../docs/optimization_guide/dldt_deployment_optimization_guide.md).

### Device
To set which device benchmarking runs on, use the `-d <device>` argument. This will tell benchmark_app to run benchmarking on that specific device. The benchmark app supports "CPU", "GPU", and "MYRIAD" (also known as [VPU](../../../docs/OV_Runtime_UG/supported_plugins/VPU.md)) devices. In order to use the GPU or VPU, the system must have the appropriate drivers installed. If no device is specified, benchmark_app will default to using CPU.

For example, to run benchmarking on GPU, use:

```
./benchmark_app -m model.xml -d GPU
```

> You may also specify "AUTO" as the device, and benchmark_app will automatically select the best device to run benchmarking on. For more information, see the [Automatic device selection](../../../docs/OV_Runtime_UG/auto_device_selection.md) page.

Reviewer 3: While the model is being loaded on GPU/VPU, inference will start on CPU. This may impact reported performance, so would be good to add here.

Reviewer: done

> You may also specify "AUTO" as the device, and benchmark_app will automatically select the best device to run benchmarking on. For more information, see the [Automatic device selection](../../../docs/OV_Runtime_UG/auto_device_selection.md) page.

Reviewer: ```suggestion
You may also specify "AUTO" as the device, in which case the benchmark_app will automatically select the best device for benchmarking and use the CPU for the time the the model is loaded on the target device. This may result in increased performance. For more information, see the [Automatic device selection](../../../docs/OV_Runtime_UG/auto_device_selection.md) page.
```

(Note: If the latency or throughput hint is set, it will automatically configure streams and batch sizes for optimal performance based on the specified device.)

### Number of iterations
By default, the benchmarking app will run for a predefined duration, repeatedly performing inferencing with the model and measuring the resulting inference speed. There are several options for setting the number of inference iterations:

> * Explicitly specify the number of iterations the model runs using the `-niter <number_of_iterations>` option
> * Set how much time the app runs for using the `-t <seconds>` option
> * Set both of them (execution will continue until both conditions are met)
> * If neither -niter nor -t are specified, the app will run for a predefined duration that depends on the device

Reviewer: ```suggestion
* Explicitly specify the number of iterations the model runs using the `-niter <number_of_iterations>` option.
* Set how much time the app runs for using the `-t <seconds>` option.
* Set both of them (execution will continue until both conditions are met).
* If neither -niter nor -t are specified, the app will run for a predefined duration that depends on the device.
```

> * Explicitly specify the number of iterations the model runs using the `-niter <number_of_iterations>` option
> * Set how much time the app runs for using the `-t <seconds>` option
> * Set both of them (execution will continue until both conditions are met)
> * If neither -niter nor -t are specified, the app will run for a predefined duration that depends on the device

Reviewer: ```suggestion
* Explicitly specify the number of iterations the model runs using the `-niter <number_of_iterations>` option.
* Set how much time the app runs for using the `-t <seconds>` option.
* Set both of them (execution will continue until both conditions are met).
* If neither -niter nor -t are specified, the app will run for a predefined duration that depends on the device.
```
